boards/qn9080dk/board.c

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (C) 2020 iosabi
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     boards_qn9080dk
+ * @{
+ *
+ * @file
+ * @brief       Board specific implementations for the QN9080DK base board
+ *
+ * @author      iosabi <iosabi@protonmail.com>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "board.h"
+
+#include "periph/gpio.h"
+
+#include "vendor/drivers/fsl_clock.h"
+
+void clocks_init(void);
+
+void board_init(void)
+{
+    /* Initialize the clock sources. */
+    clocks_init();
+
+    /* Initialize LEDs and Buttons. */
+    gpio_init(LED_RED_PIN, GPIO_OUT);
+    gpio_init(LED_GREEN_PIN, GPIO_OUT);
+    gpio_init(LED_BLUE_PIN, GPIO_OUT);
+    gpio_init(BTN1_PIN, BTN1_MODE);
+    gpio_init(BTN2_PIN, BTN2_MODE);
+
+    /* initialize the CPU */
+    cpu_init();
+}
+
+void clocks_init(void)
+{
+    /* Set up clock selectors - Attach clocks to the peripheries */
+
+    /* Switch XTAL_CLK to 32M */
+    CLOCK_AttachClk(k32M_to_XTAL_CLK);
+    /* Switch 32K_CLK to XTAL32K */
+    CLOCK_AttachClk(kXTAL32K_to_32K_CLK);
+    /* Switch SYS_CLK to XTAL */
+    CLOCK_AttachClk(kXTAL_to_SYS_CLK);
+    /* Switch WDT_CLK to APB */
+    CLOCK_AttachClk(kAPB_to_WDT_CLK);
+
+    /* Set up dividers */
+
+    /* Set OSC32M_DIV divider to value 2 */
+    CLOCK_SetClkDiv(kCLOCK_DivOsc32mClk, 1U);
+    /* Set XTAL_DIV divider to value 2 */
+    CLOCK_SetClkDiv(kCLOCK_DivXtalClk, 1U);
+    /* Set AHB_DIV divider to value 2 */
+    CLOCK_SetClkDiv(kCLOCK_DivAhbClk, 1U);
+    /* Set FRG_MULT1 to value 0, Set FRG_DIV1 to value 255 */
+    CLOCK_SetClkDiv(kCLOCK_DivFrg1, 0U);
+    /* Set FRG_MULT0 to value 0, Set FRG_DIV0 to value 255 */
+    CLOCK_SetClkDiv(kCLOCK_DivFrg0, 0U);
+    /* Set APB_DIV divider to value 1 */
+    CLOCK_SetClkDiv(kCLOCK_DivApbClk, 0U);
+}
```

Comment:
ok, the thinking here is that whether you have an external XTAL for 32 MHz (or 16 MHz) and another external xtal for the 32.768 KHz depends on the board wiring, you can have a design that uses only the internal ones (32MHz and 32.768 KHz), then based on this you would have different clock setup for the board which also changes how do you configure the APB bus clock for the peripherals.

Should all of this be done using macros in `cpu_conf.h` that the board can override, but the code lives in cpu_init() (or a function called from there)?

Reply:
> Should all of this be done using macros in cpu_conf.h that the board can override, but the code lives in cpu_init()

Yes that's the idea.
It's better to have all the logic in one place and not spread out / copied across multiple boards.

You could have a `BOARD_HAS_XOSC32K` define that indicates if a 32kHz crystal is present, and another `BOARD_XOSC_FREQ` / `BOARD_HAS_XOSC` define to indicate the frequency / presence of a high frequency oscillator.

There are no naming conventions for this, just use whatever best matches the description in the data sheet.

Reply:
I moved this to cpu.c and added a whole lot of config.

I'm not entirely sure how is the Kconfig supposed to work. If I run `make BOARD=qn9080dk -C examples/hello-world/` the `autoconf.h` is not generated and the CONFIG_ variables with the defaults in the Kconfig are not set, not even the ones the board should be forcing.

I ended up adding the same logic with CONFIG_* variables in the cpu_conf.h, maybe this is just a transition thing but it would be better if I don't have to do that. On the other hand adding them to cpu_conf.h lets me document them so doxygen can pick them up.

Reply:
Yes the Kconfig transition is not yet complete, not all modules can be selected in Kconfig yet, so it's not enabled by default.

I *think* you don't need to model the board oscillator config in Kconfig since it's soldered in place.

Reply:
I added the matching "CONFIG_*" to the CFLAGS in `Makefile.include` in the board, so the if you don't have an autoconf.h you still get the CONFIG_BOARD_HAS_XTAL_32M and related macros defined.